### PR TITLE
Add data structure for storing attach detach controller state.

### PR DESCRIPTION
**Pull request by saad-ali**
This PR introduces the data structure for maintaining the in-memory state for the new attach/detach controller (#20262).

### CODE REVIEW

File: pkg/controller/volume/cache/attach_detach_volume_cache.go [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+// an error is returned.
+// If the node already exists for the specified volume, this is a no-op.
+func (vc *attachDetachVolumeCache) AddNode(nodeName, volumeName string) error {
+	vc.Lock()
+	defer vc.Unlock()
+
+	vol, volExists := vc.volumesManaged[volumeName]
+	if !volExists {
+		return fmt.Errorf(
+			"failed to add pod %q to node %q volume %q--no volume with that name exists in the list of managed volumes",
```
Comment (swagiaal) [inline pkg/controller/volume/cache/attach_detach_volume_cache.go]: should be "failed to add node X to volume Y" I think

File: pkg/controller/volume/cache/attach_detach_volume_cache.go [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+	defer vc.Unlock()
+
+	volObj, volExists := vc.volumesManaged[volumeName]
+	if !volExists {
+		return fmt.Errorf(
+			"failed to delete volume %q--no volume with that name exists in the list of managed volumes",
+			volumeName)
+	}
+
+	if len(volObj.attachedNodes) > 0 {
```
Comment (swagiaal) [inline pkg/controller/volume/cache/attach_detach_volume_cache.go]: this behavior should be added to the function's comment

File: pkg/controller/volume/cache/attach_detach_volume_cache.go [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+	nodeObj, nodeExists := volObj.attachedNodes[nodeName]
+	if !nodeExists {
+		return fmt.Errorf(
+			"failed to delete node %q from volume %q--no node with the that name exists in the list of attached nodes for that volume",
+			nodeName,
+			volumeName)
+	}
+
+	if len(nodeObj.scheduledPods) > 0 {
```
Comment (swagiaal) [inline pkg/controller/volume/cache/attach_detach_volume_cache.go]: same as above.

**Comment by swagiaal** [comment]
A few nits but otherwise LGTM

**Comment by saad-ali** [comment]
Thanks @swagiaal. PTAL

**Pull request merged**